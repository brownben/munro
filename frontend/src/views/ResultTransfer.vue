<!--
  Result Transfer

  Change the competitor the result is credited to
-->

<template>
  <div>
    <h1>Transfer Result</h1>
    <form @submit.prevent="transfer()">
      <label>League:</label>
      <dropdown-input
        v-model="league"
        :list="leagues.map(league => league.name)"
        :include-blank="true"
      />
      <label>Event:</label>
      <dropdown-input
        v-model="event"
        :list="eventsInLeague.map(event => event.name + ' - ' + event.date)"
        :include-blank="true"
      />
      <label>Course:</label>
      <dropdown-input v-model="course" :list="coursesInLeague" :include-blank="true" />
      <label>Results:</label>
      <dropdown-input
        v-model="result"
        :list="resultsForEvent.map(result => result.position + ' - ' + elapsedTime(result.time) + ' (' + result.name + ')')"
        :include-blank="true"
      />
      <label>Competitor:</label>
      <dropdown-input
        v-model="competitor"
        :list="competitorsForLeague.map(competitorTransformForSelect)"
        :include-blank="true"
      />
      <button>Transfer Result</button>
    </form>
  </div>
</template>

<script>
import axios from 'axios'
import DropdownInput from '@/components/DropdownInput'

export default {
  components: {
    DropdownInput: DropdownInput,
  },

  data: () => ({
    competitors: [],
    leagues: [],
    events: [],
    results: [],
    league: '',
    event: '',
    course: '',
    competitor: '',
    result: '',
  }),

  computed: {
    competitorsForLeague: function () {
      const filtered = this.competitors.filter(competitor => competitor.league === this.league && competitor.course === this.course)
      return filtered.sort((a, b) => a.name > b.name)
    },

    coursesInLeague: function () {
      const selectedLeague = this.leagues.filter(league => league.name === this.league)
      if (selectedLeague.length > 0) return selectedLeague[0].courses
      else return []
    },

    eventsInLeague: function () {
      return this.events.filter(event => event.league === this.league)
    },

    resultsForEvent: function () {
      let event = ''
      const selectedEvent = this.events.filter(event => event.name === this.event.split(' - ')[0] && event.date === this.event.split(' - ')[1])
      if (selectedEvent.length > 0) event = selectedEvent[0].id
      return this.results
        .filter(result => result.event === event && result.course === this.course)
        .sort((a, b) => {
          if (parseInt(a.position) === parseInt(b.position)) return 0
          else if (a.position === null || a.position === undefined) return 1
          else if (b.position === null || b.position === undefined) return -1
          else if (parseInt(a.position) > parseInt(b.position)) return 1
          else return -1
        })
    },
  },

  created: async function () {
    await this.getCompetitors()
    await this.getLeagues()
    await this.getEvents()
    await this.getResults()
  },

  methods: {
    getCompetitors: function () {
      return axios.get('/api/competitors')
        .then(response => { this.competitors = response.data })
        .catch(() => this.$messages.addMessage('Problem Fetching Competitors'))
    },

    getLeagues: function () {
      return axios.get('/api/leagues')
        .then(response => { this.leagues = response.data })
        .catch(() => this.$messages.addMessage('Problem Fetching Leagues'))
    },

    getEvents: function () {
      return axios.get('/api/events')
        .then(response => { this.events = response.data })
        .catch(() => this.$messages.addMessage('Problem Fetching Events'))
    },

    getResults: function () {
      return axios.get('/api/results')
        .then(response => { this.results = response.data })
        .catch(() => this.$messages.addMessage('Problem Fetching Results'))
    },

    validateForm: function () {
      return (this.event !== '' && this.course !== '' && this.result !== '')
    },

    twoDigits: function (number) {
      if (number.toString().length < 2) return '0' + number.toString()
      else return number
    },

    elapsedTime: function (totalTimeInSeconds) {
      const timeInMinutes = Math.floor(totalTimeInSeconds / 60)
      const timeInSeconds = Math.abs(totalTimeInSeconds % 60)
      return this.twoDigits(timeInMinutes) + ':' + this.twoDigits(timeInSeconds)
    },

    elapsedTimeToSeconds: function (time) {
      return parseInt(time.split(':')[0]) * 60 + parseInt(time.split(':')[1])
    },

    competitorTransformForSelect: function (competitor) {
      if (competitor.club && competitor.ageClass) return competitor.name + ' (' + competitor.ageClass + ', ' + competitor.club + ') [' + competitor.id + ']'
      else if (competitor.club) return competitor.name + ' (' + competitor.club + ') [' + competitor.id + ']'
      else if (competitor.ageClass) return competitor.name + ' (' + competitor.ageClass + ') [' + competitor.id + ']'
      else return competitor.name + ' [' + competitor.id + ']'
    },

    transfer: function () {
      if (this.validateForm()) {
        let event = ''
        const selectedEvent = this.events.find(event => event.name === this.event.split(' - ')[0] && event.date === this.event.split(' - ')[1])
        if (selectedEvent) event = selectedEvent.id
        const competitor = this.competitor.replace(/.*\[|\]/g, '')
        const result = this.results.find(result =>
          result.course === this.course &&
          result.event === event &&
          result.time === this.elapsedTimeToSeconds(this.result.match(/-.*\(/)[0].slice(2, -2))
        )
        return axios.post('/api/results/transfer', {
          competitor: competitor,
          result: result.id,
        })
          .then(response => this.returnToCompetitorsPage(response))
          .catch(error => this.$messages.addMessage(error.response.data.message))
      }
      else this.$messages.addMessage('The Competitors Selected are the Same')
    },

    returnToCompetitorsPage: function (response) {
      // Go to league page after successful update/ creation
      this.$messages.addMessage(response.data.message)
      this.$router.push('/competitors')
    },
  },
}
</script>
<style lang="stylus" scoped>
@import '../assets/styles/helpers.styl'
@import '../assets/styles/inputs.styl'

h1
  margin-bottom: 0.5rem
</style>
