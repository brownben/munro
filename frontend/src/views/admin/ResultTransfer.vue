<template>
  <Layout title="Transfer Result">
    <Meta
      :head="{
        meta: { name: 'robots', content: 'noindex' },
      }"
      title="Munro - Transfer Results"
      description
    />

    <form class="col-span-2" @submit.prevent="transferResult">
      <DropdownInput
        v-model="choices.league"
        :list="leagues.map((league) => league.name)"
        :include-blank="true"
        label="League:"
      />
      <DropdownInput
        v-model="choices.event"
        :listWithDifferentValue="
          eventsInLeague.map((event) => ({
            value: event.id,
            text: `${event.name} (${event.date})`,
          }))
        "
        :include-blank="true"
        label="Event:"
        class="mt-4"
      />
      <DropdownInput
        v-model="choices.course"
        :list="courses"
        :include-blank="true"
        label="Course"
        class="mt-4"
      />
      <DropdownInput
        v-model="choices.result"
        :listWithDifferentValue="
          resultsInEvent?.map((result) => ({
            text: `${result.position} - ${elapsedTime(result.time)} (${
              result.name
            })`,
            value: result.id.toString(),
          })) ?? []
        "
        :include-blank="true"
        label="Result:"
        class="mt-4"
      />
      <DropdownInput
        v-model="choices.competitor"
        :listWithDifferentValue="
          competitorsInLeague.map((competitor) => ({
            value: competitor.id.toString(),
            text: competitorToText(competitor),
          }))
        "
        :include-blank="true"
        label="Competitor:"
        class="mt-4"
      />

      <button class="mt-8 button-lg">Transfer Result</button>
    </form>
  </Layout>
</template>

<script lang="ts" setup>
import { ref, onMounted, computed } from 'vue'
import { useRouter, useRoute } from 'vue-router'
import { useStore } from 'vuex'

import Layout from '../../components/Layout.vue'
import DropdownInput from '../../components/inputs/DropdownInput.vue'

import { getLeagues } from '../../api/leagues'
import { getEvents } from '../../api/events'
import { getCompetitors } from '../../api/competitors'
import {
  getResults,
  transferResult as apiTransferResult,
} from '../../api/results'

import { sortEventResults, SortablePropertiesEvent } from '../../scripts/sort'
import { elapsedTime } from '../../scripts/time'

const store = useStore()
const router = useRouter()
const route = useRoute()

const leagues = ref<League[]>([])
const events = ref<Event[]>([])
const competitors = ref<Competitor[]>([])
const results = ref<EventResult[]>([])
const choices = ref({
  league: '',
  event: '',
  course: '',
  competitor: '',
  result: '',
})

onMounted(async () => {
  leagues.value = (await getLeagues()) ?? []
  events.value = (await getEvents()) ?? []
  results.value = (await getResults()) ?? []
  competitors.value = (await getCompetitors()) ?? []
})

const courses = computed(
  () =>
    leagues.value?.find((league) => league.name === choices.value.league)
      ?.courses ?? []
)
const eventsInLeague = computed(() =>
  events.value?.filter((event) => event.league === choices.value.league)
)
const competitorsInLeague = computed(() =>
  competitors.value
    ?.filter(
      (competitor) =>
        competitor.league === choices.value.league &&
        competitor.course === choices.value.course
    )
    ?.sort((a, b) => (a.name > b.name ? 1 : -1))
)
const resultsInEvent = computed(() =>
  results.value
    .filter(
      (result) =>
        result.event === choices.value.event &&
        result.course === choices.value.course
    )
    .sort(
      sortEventResults({
        ascending: false,
        by: SortablePropertiesEvent.position,
      })
    )
)

const competitorToText = (competitor: Competitor) => {
  if (competitor.club && competitor.ageClass)
    return `${competitor.name} (${competitor.ageClass}, ${competitor.club}) [${competitor.id}]`
  else if (competitor.club)
    return `${competitor.name} (${competitor.club}) [${competitor.id}]`
  else if (competitor.ageClass)
    return `${competitor.name} (${competitor.ageClass}) [${competitor.id}]`
  else return `${competitor.name} [${competitor.id}]`
}

const validateForm = () => {
  if (
    choices.value.event !== '' &&
    choices.value.course !== '' &&
    choices.value.result !== '' &&
    choices.value.competitor !== ''
  )
    return true
  else {
    store.dispatch(
      'createMessage',
      'Please Ensure Result and Competitor Fields are not Blank'
    )
    return false
  }
}

const transferResult = () =>
  apiTransferResult({
    competitor: choices.value.competitor,
    result: choices.value.result,
  })
    .then(() => router.push(`/results/${choices.value.result}`))
    .catch(() => false)
</script>
