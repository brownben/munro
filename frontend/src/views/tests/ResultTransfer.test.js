import { mount } from '@vue/test-utils'
import ResultTransfer from '@/views/ResultTransfer'

test('Is a Vue Instance', () => {
  const wrapper = mount(ResultTransfer)
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly', () => {
  const wrapper = mount(ResultTransfer)
  expect(wrapper.element).toMatchSnapshot()
})

