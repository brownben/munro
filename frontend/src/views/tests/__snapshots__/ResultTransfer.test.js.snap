// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly 1`] = `
<div>
  <h1>
    Transfer Result
  </h1>
   
  <form>
    <label>
      League:
    </label>
     
    <div
      class="dropdown-input"
    >
      <select>
        <option />
         
      </select>
       
      <label>
        <svg
          fill="#9E9E9E"
          height="24"
          viewBox="0 0 24 24"
          width="24"
        >
          <path
            d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
          />
           
          <path
            d="M0-.75h24v24H0z"
            fill="none"
          />
        </svg>
      </label>
    </div>
     
    <label>
      Event:
    </label>
     
    <div
      class="dropdown-input"
    >
      <select>
        <option />
         
      </select>
       
      <label>
        <svg
          fill="#9E9E9E"
          height="24"
          viewBox="0 0 24 24"
          width="24"
        >
          <path
            d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
          />
           
          <path
            d="M0-.75h24v24H0z"
            fill="none"
          />
        </svg>
      </label>
    </div>
     
    <label>
      Course:
    </label>
     
    <div
      class="dropdown-input"
    >
      <select>
        <option />
         
      </select>
       
      <label>
        <svg
          fill="#9E9E9E"
          height="24"
          viewBox="0 0 24 24"
          width="24"
        >
          <path
            d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
          />
           
          <path
            d="M0-.75h24v24H0z"
            fill="none"
          />
        </svg>
      </label>
    </div>
     
    <label>
      Results:
    </label>
     
    <div
      class="dropdown-input"
    >
      <select>
        <option />
         
      </select>
       
      <label>
        <svg
          fill="#9E9E9E"
          height="24"
          viewBox="0 0 24 24"
          width="24"
        >
          <path
            d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
          />
           
          <path
            d="M0-.75h24v24H0z"
            fill="none"
          />
        </svg>
      </label>
    </div>
     
    <label>
      Competitor:
    </label>
     
    <div
      class="dropdown-input"
    >
      <select>
        <option />
         
      </select>
       
      <label>
        <svg
          fill="#9E9E9E"
          height="24"
          viewBox="0 0 24 24"
          width="24"
        >
          <path
            d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
          />
           
          <path
            d="M0-.75h24v24H0z"
            fill="none"
          />
        </svg>
      </label>
    </div>
     
    <button>
      Transfer Result
    </button>
  </form>
</div>
`;
